File: dotnet/Allmark/Inline/EmphasisRule.cs
namespace Allmark.Inline;

using Allmark.Types;

public static class EmphasisRule
{
	public static InlineRule Create()
	{
		return new InlineRule
		{
			Name = "emphasis",
			Test = TestEmphasis,
		};
	}

	private static bool TestEmphasis(InlineParserState state, MarkdownNode parent)
	{
		var ch = Utils.GetChar(state.Src, state.I);
		if ((ch == '*' || ch == '_') && !Utils.IsEscaped(state.Src, state.I))
		{
			var start = state.I;
			var end = state.I;

			// Get the markup
			var markup = ch.ToString();
			for (var j = start + 1; j < state.Src.Length; j++)
			{
				if (state.Src[j] == ch)
				{
					markup += ch;
					end++;
				}
				else
				{
					break;
				}
			}

			// TODO: Better space checks including start/end of line
			var codeBefore = start > 0 ? state.Src[start - 1] : default;
			var spaceBefore = start == 0 || Utils.IsUnicodeSpace(codeBefore);
			var punctuationBefore = !spaceBefore && Utils.IsUnicodePunctuation(codeBefore);

			var codeAfter = end + 1 < state.Src.Length ? state.Src[end + 1] : default;
			var spaceAfter = end == state.Src.Length - 1 || Utils.IsUnicodeSpace(codeAfter);
			var punctuationAfter = !spaceAfter && Utils.IsUnicodePunctuation(codeAfter);

			// "A left-flanking delimiter run is a delimiter run that is (1) not
			// followed by Unicode whitespace, and either (2a) not followed by a
			// punctuation character, or (2b) followed by a punctuation character
			// and preceded by Unicode whitespace or a punctuation character. For
			// purposes of this definition, the beginning and the end of the line
			// count as Unicode whitespace."
			var leftFlanking =
				!spaceAfter &&
				(!punctuationAfter || (punctuationAfter && (spaceBefore || punctuationBefore)));

			// "A right-flanking delimiter run is a delimiter run that is (1) not
			// preceded by Unicode whitespace, and either (2a) not preceded by a
			// punctuation character, or (2b) preceded by a punctuation character
			// and followed by Unicode whitespace or a punctuation character. For
			// purposes of this definition, the beginning and the end of the line
			// count as Unicode whitespace"
			var rightFlanking =
				!spaceBefore &&
				(!punctuationBefore || (punctuationBefore && (spaceAfter || punctuationAfter)));

			// TODO: Precedence
			// Loop backwards through delimiters to find a matching one that does
			// not take precedence, and ideally has the same length
			Delimiter? startDelimiter = null;
			var startIndex = -1;
			var i = state.Delimiters.Count;
			while (i-- > 0)
			{
				var prevDelimiter = state.Delimiters[i];
				if (!prevDelimiter.Handled)
				{
					if (prevDelimiter.Markup == ch.ToString())
					{
						if (prevDelimiter.Length == markup.Length)
						{
							startDelimiter = prevDelimiter;
							startIndex = i;
							break;
						}
						else if (startDelimiter == null)
						{
							startDelimiter = prevDelimiter;
							startIndex = i;
						}
					}
					else if (prevDelimiter.Markup == "*" || prevDelimiter.Markup == "_")
					{
						continue;
					}
					else
					{
						break;
					}
				}
			}

			// Check if it's a closing emphasis
			if (startDelimiter != null)
			{
				var canClose =
					(rightFlanking ||
						// Check if it's a continuing part of a three-run delimiter
						(state.I > 0 && state.Src[state.I - 1] == ch)) &&
					startDelimiter.Markup == ch.ToString() &&
					// "Emphasis with _ is not allowed inside words"
					(ch != '_' || spaceAfter || punctuationAfter) &&
					// "[A] delimiter that can both open and close ... cannot form
					// emphasis if the sum of the lengths of the delimiter runs
					// containing the opening and closing delimiters is a multiple
					// of 3 unless both lengths are multiples of 3."
					(!leftFlanking ||
						(markup.Length + startDelimiter.Length) % 3 != 0 ||
						(markup.Length % 3 == 0 && startDelimiter.Length % 3 == 0));
				if (canClose)
				{
					// Convert the text node into an emphasis node with a new text child
					// followed by the other children of the parent (if any)
					var j = parent.Children?.Count ?? 0;
					while (j-- > 0)
					{
						var lastNode = parent.Children?[j];
						if (lastNode?.Index == startDelimiter.Start)
						{
							// If it's longer than the last delimiter, or longer
							// than two, save some for the next go-round
							markup = markup.Substring(0, Math.Min(markup.Length, Math.Min(startDelimiter.Length, 2)));

							var text = Utils.NewNode("text", false, lastNode.Index, lastNode.Line, 1, ch.ToString(), 0);
							text.Markup = lastNode.Markup.Substring(startDelimiter.Length) ?? "";

							var movedNodes = parent.Children!.Skip(j + 1).ToList() ?? [];
							parent.Children!.RemoveRange(j + 1, movedNodes.Count);

							if (markup.Length < startDelimiter.Length)
							{
								lastNode.Markup = lastNode.Markup.Substring(0, startDelimiter.Length - markup.Length);
								var emphasis = Utils.NewNode(
									markup.Length == 2 ? "strong" : "emphasis",
									false,
									lastNode.Index + markup.Length,
									lastNode.Line,
									1,
									markup,
									0,
									[text, .. movedNodes]);
								parent.Children!.Add(emphasis);
							}
							else
							{
								lastNode.Type = markup.Length == 2 ? "strong" : "emphasis";
								lastNode.Markup = markup;
								lastNode.Children = [text, .. movedNodes];
							}

							state.I += markup.Length;

							// Mark delimiters between the start and end as handled,
							// as they can't start anything anymore
							var d = state.Delimiters.Count;
							while (d-- > 0)
							{
								if (d == startIndex)
								{
									break;
								}
								var prevDelimiter = state.Delimiters[d];
								prevDelimiter.Handled = true;
							}

							// Mark the start delimiter handled if all its chars are used up
							startDelimiter.Length -= markup.Length;
							if (startDelimiter.Length == 0)
							{
								startDelimiter.Handled = true;
							}

							return true;
						}
					}
				}
			}

			// Check if it's an opening emphasis
			var canOpen =
				leftFlanking &&
				// "Emphasis with _ is not allowed inside words"
				(ch != '_' || spaceBefore || punctuationBefore);
			if (canOpen)
			{
				// Add a new text node which may turn into emphasis
				var text = Utils.NewNode("text", false, start, state.Line, 1, markup, 0);
				parent.Children!.Add(text);

				state.I += markup.Length;
				state.Delimiters.Add(new Delimiter { Markup = ch.ToString(), Start = start, Length = markup.Length });

				return true;
			}

			Utils.AddMarkupAsText(markup, state, parent);

			return true;
		}

		return false;
	}
}

File: web/src/inline/emphasisRule.ts
import type Delimiter from "../types/Delimiter";
import type InlineParserState from "../types/InlineParserState";
import type InlineRule from "../types/InlineRule";
import type MarkdownNode from "../types/MarkdownNode";
import addMarkupAsText from "../utils/addMarkupAsText";
import { ASTERISK_CODE, UNDERSCORE_CODE } from "../utils/charCodes";
import isUnicodePunctuation from "../utils/isUnicodePunctuation";
import isUnicodeSpace from "../utils/isUnicodeSpace";
import newInline from "../utils/newInline";
import newText from "../utils/newText";
import { appendChild, getLastDescendant, spliceTextNode } from "../utils/nodeUtils";

const rule: InlineRule = {
	name: "emphasis",
	test: testEmphasis,
	precedence: 10,
};
export default rule;

function testEmphasis(state: InlineParserState, parent: MarkdownNode): boolean {
	let charCode = state.src.charCodeAt(state.i);
	if (!state.isEscaped && (charCode === ASTERISK_CODE || charCode === UNDERSCORE_CODE)) {
		let char = state.src[state.i];
		let start = state.i;
		let end = state.i;

		// Get the markup
		let markup = char;
		for (let i = start + 1; i < state.src.length; i++) {
			if (state.src.charCodeAt(i) === charCode) {
				markup += char;
				end++;
			} else {
				break;
			}
		}
		//let markup = state.src.substring(start, end);

		// TODO: Better space checks including start/end of line
		let codeBefore = state.src.charCodeAt(start - 1);
		let spaceBefore = start === 0 || isUnicodeSpace(codeBefore);
		let punctuationBefore = !spaceBefore && isUnicodePunctuation(codeBefore);

		let codeAfter = state.src.charCodeAt(end + 1);
		let spaceAfter = end === state.src.length - 1 || isUnicodeSpace(codeAfter);
		let punctuationAfter = !spaceAfter && isUnicodePunctuation(codeAfter);

		// "A left-flanking delimiter run is a delimiter run that is (1) not
		// followed by Unicode whitespace, and either (2a) not followed by a
		// punctuation character, or (2b) followed by a punctuation character
		// and preceded by Unicode whitespace or a punctuation character. For
		// purposes of this definition, the beginning and the end of the line
		// count as Unicode whitespace."
		let leftFlanking =
			!spaceAfter &&
			(!punctuationAfter || (punctuationAfter && (spaceBefore || punctuationBefore)));

		// "A right-flanking delimiter run is a delimiter run that is (1) not
		// preceded by Unicode whitespace, and either (2a) not preceded by a
		// punctuation character, or (2b) preceded by a punctuation character
		// and followed by Unicode whitespace or a punctuation character. For
		// purposes of this definition, the beginning and the end of the line
		// count as Unicode whitespace"
		let rightFlanking =
			!spaceBefore &&
			(!punctuationBefore || (punctuationBefore && (spaceAfter || punctuationAfter)));

		// Loop backwards through delimiters to find a matching one that does
		// not take precedence, and ideally has the same length
		let startDelimiter: Delimiter | undefined;
		let startIndex = -1;
		let i = state.delimiters.length;
		while (i--) {
			let prevDelimiter = state.delimiters[i];
			if (!prevDelimiter.handled) {
				if (prevDelimiter.markup === char) {
					if (prevDelimiter.length === markup.length) {
						startDelimiter = prevDelimiter;
						startIndex = i;
						break;
					} else if (!startDelimiter) {
						startDelimiter = prevDelimiter;
						startIndex = i;
					}
				} else if ((prevDelimiter.precedence ?? 0) <= rule.precedence!) {
					// Same or lower precedence delimiters can be skipped over
					continue;
				} else {
					// Higher precedence delimiters block
					break;
				}
			}
		}

		// Check if it's a closing emphasis
		if (startDelimiter !== undefined) {
			let canClose =
				(rightFlanking ||
					// Check if it's a continuing part of a three-run delimiter
					state.src.charCodeAt(state.i - 1) === charCode) &&
				startDelimiter.markup === char &&
				// "Emphasis with _ is not allowed inside words"
				(charCode !== UNDERSCORE_CODE || spaceAfter || punctuationAfter) &&
				// "[A] delimiter that can both open and close ... cannot form
				// emphasis if the sum of the lengths of the delimiter runs
				// containing the opening and closing delimiters is a multiple
				// of 3 unless both lengths are multiples of 3."
				(!leftFlanking ||
					(markup.length + startDelimiter.length) % 3 !== 0 ||
					(markup.length % 3 === 0 && startDelimiter.length % 3 === 0));
			if (canClose) {
				// Convert the text node into an emphasis node with a new text child
				// followed by the other children of the parent (if any)
				let lastDescendant = getLastDescendant(parent)!;
				let lastNode = lastDescendant;
				while (lastNode !== parent) {
					if (
						lastNode.depth === parent.depth + 1 &&
						lastNode.index === state.parentIndex + startDelimiter.start
					) {
						// If it's longer than the last delimiter, or longer
						// than two, save some for the next go-round
						markup = markup.substring(0, Math.min(startDelimiter.length, 2));

						let content = lastNode.content.slice(startDelimiter.length);
						let text = newText(lastNode.index, lastNode.line, content, 0);
						text.length = text.content.length;

						if (markup.length < startDelimiter.length) {
							let remainingStart = lastNode.index + startDelimiter.length - markup.length;
							lastNode.content = lastNode.content.substring(
								0,
								startDelimiter.length - markup.length,
							);
							lastNode.length = lastNode.content.length;
							let emphasis = newInline(
								markup.length === 2 ? "strong" : "emphasis",
								remainingStart,
								lastNode.line,
								markup,
								0,
							);
							emphasis.length = state.parentIndex + state.i - remainingStart + markup.length;

							text.previousNode = emphasis;
							text.nextNode = lastNode.nextNode;
							emphasis.previousNode = lastNode;
							emphasis.nextNode = text;
							lastNode.nextNode = emphasis;

							let nextNode = lastNode;
							while (nextNode !== lastDescendant) {
								nextNode.depth++;
								nextNode = nextNode.nextNode!;
							}
							lastDescendant.depth++;

							lastNode.depth = parent.depth + 1;
							emphasis.depth = parent.depth + 1;
							text.depth = emphasis.depth + 1;
						} else {
							lastNode.type = markup.length === 2 ? "strong" : "emphasis";
							lastNode.markup = markup;
							spliceTextNode(parent, text, lastNode, lastDescendant);
							lastNode.length = state.parentIndex + state.i - lastNode.index + markup.length;
						}

						state.i += markup.length;

						// Mark delimiters between the start and end as handled,
						// as they can't start anything anymore
						let d = state.delimiters.length;
						while (d--) {
							if (d === startIndex) {
								break;
							}
							let prevDelimiter = state.delimiters[d];
							prevDelimiter.handled = true;
						}

						// Mark the start delimiter handled if all its chars are used up
						startDelimiter.length -= markup.length;
						if (!startDelimiter.length) {
							startDelimiter.handled = true;
						}

						return true;
					}

					lastNode = lastNode.previousNode!;
				}
			}
		}

		// Check if it's an opening emphasis
		let canOpen =
			leftFlanking &&
			// "Emphasis with _ is not allowed inside words"
			(charCode !== UNDERSCORE_CODE || spaceBefore || punctuationBefore);
		if (canOpen) {
			// Add a new text node which may turn into emphasis
			let text = newText(state.parentIndex + start, state.line, markup, 0);
			appendChild(parent, text);

			state.i += markup.length;
			state.delimiters.push({
				markup: char,
				start,
				length: markup.length,
				precedence: rule.precedence,
			});

			return true;
		}

		addMarkupAsText(markup, state, parent);

		return true;
	}

	return false;
}
